src/logging.c

```diff
@@ -626,20 +626,56 @@ static void parse_and_create_handlers_list(void)
   }
 }
 
+rcutils_ret_t rcutils_logging_allocator_initialize(
+  const rcutils_allocator_t * allocator)
+{
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    allocator, "invalid allocator", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (rcutils_allocator_is_valid(&g_rcutils_logging_allocator)) {
+    return RCUTILS_RET_OK;
+  }
+  g_rcutils_logging_allocator = *allocator;
+
+  return RCUTILS_RET_OK;
+}
+
 rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allocator)
 {
   if (g_rcutils_logging_initialized) {
     return RCUTILS_RET_OK;
   }
 
-  if (!rcutils_allocator_is_valid(&allocator)) {
-    RCUTILS_SET_ERROR_MSG("Provided allocator is invalid.");
+  if (rcutils_logging_allocator_initialize(&allocator) != RCUTILS_RET_OK) {
     return RCUTILS_RET_INVALID_ARGUMENT;
   }
-  g_rcutils_logging_allocator = allocator;
 
   g_rcutils_logging_output_handler = &rcutils_logging_console_output_handler;
-  g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
+  // Check for the environment variable for default logger level
+  const char * env_default_level;
+  const char * ret_str_level =
+    rcutils_get_env("RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL", &env_default_level);
+  if (NULL != ret_str_level) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting environment variable RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL: %s", ret_str_level);
+    g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
+  } else if (strcmp(env_default_level, "") == 0) {
+    // Environment variable is empty, use default
+    g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
+  } else {
+    // Try to convert the environment variable value to a severity level
+    int env_severity;
+    rcutils_ret_t severity_ret = rcutils_logging_severity_level_from_string(
+      env_default_level, g_rcutils_logging_allocator, &env_severity);
+    if (severity_ret == RCUTILS_RET_OK) {
+      g_rcutils_logging_default_logger_level = env_severity;
+    } else {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Invalid severity level '%s' in RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL. Using default.",
+        env_default_level);
+      g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
+    }
+  }
```

Comment:
This is new behavior we introduce with this PR.

before, the default global log level is fixed as RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL statically, that leads the situation if the initialization of context is not finished, there is no way to print the debug log message unless rebuild the system with changing RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL.
but now, user application can set `RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL` environmental variable to change the default global log level at runtime without any code change. this gives more flexibility to debug the initialization process for ROS 2 system libraries.